onewire/onewire.go

```diff
@@ -0,0 +1,213 @@
+// Package wire implements the Dallas Semiconductor Corp.'s 1-wire bus system.
+//
+// Wikipedia: https://en.wikipedia.org/wiki/1-Wire
+package onewire // import "tinygo.org/x/drivers/onewire"
+
+import (
+	"errors"
+	"machine"
+	"time"
+)
+
+// OneWire ROM commands
+const (
+	READ_ROM   uint8 = 0x33
+	MATCH_ROM  uint8 = 0x55
+	SKIP_ROM   uint8 = 0xCC
+	SEARCH_ROM uint8 = 0xF0
+)
+
+// Device wraps a connection to an 1-Wire devices.
+type Device struct {
+	p machine.Pin
+}
+
+// Config wraps a configuration to an 1-Wire devices.
+type Config struct{}
+
+// Errors list
+var (
+	errNoPresence  = errors.New("Error: OneWire. No devices on the bus.")
+	errReadAddress = errors.New("Error: OneWire. Read address error: CRC mismatch.")
+)
+
+// New creates a new GPIO 1-Wire connection.
+// The pin must be pulled up to the VCC via a resistor greater than 500 ohms (default 4.7k).
+func New(p machine.Pin) Device {
+	return Device{
+		p: p,
+	}
+}
+
+// Configure initializes the protocol.
+func (d *Device) Configure(config Config) {}
+
+// Reset pull DQ line low, then up.
+func (d Device) Reset() error {
+	d.p.Configure(machine.PinConfig{Mode: machine.PinOutput})
+	time.Sleep(480 * time.Microsecond)
+	d.p.Configure(machine.PinConfig{Mode: machine.PinInput})
+	time.Sleep(70 * time.Microsecond)
+	precence := d.p.Get()
+	time.Sleep(410 * time.Microsecond)
+	if precence {
+		return errNoPresence
+	}
+	return nil
+}
+
+// WriteBit transmits a bit to 1-Wire bus.
+func (d Device) WriteBit(data uint8) {
+	d.p.Configure(machine.PinConfig{Mode: machine.PinOutput})
+	if data&1 == 1 { // Send '1'
+		time.Sleep(5 * time.Microsecond)
```

Comment:
Sorry for the late comment, but I wasn't around in January when this PR was started.

I just want to point out that `time.Sleep()` yields to the goroutine scheduler. There is no guarantee that this will sleep for 5 microseconds. It could sleep for 100s of microseconds, or several milliseconds.

Since TinyGo does not provide a non-yielding `runtime.delayMicros()` or `machine.delayMicros()` function, as far as I know, maybe there is no other option right now.

Reply:
I have conducted several experiments for the chips I have (rp2040, stm32f103, stm32f411, attiny85, atmega328p). Created a project:
```
package main

import "time"

func main() {
	for {
		time.Sleep(5 * time.Microsecond)
	}
}
```
Got an assembler listing and studied them. In all variants, a similar tick counting mechanism was used to organize the delay.

Reply:
That's interesting.. Have you looked at the SAMD21? I've been struggling with it. `time.Sleep()` seems to have an overhead of 250-400 microseconds. The following program prints: `elapsed millis: 36621` instead of a number close to `500` milliseconds:

```go
package main

import "time"

func main() {
        time.Sleep(2 * time.Second) // wait for serial monitor

        println("Starting...")
        start := time.Now()

        for i := 0; i < 100_000; i++ {
                time.Sleep(5 * time.Microsecond)
        }

        elapsed := time.Since(start)
        println("elapsed millis:", elapsed.Milliseconds())
}
```